From 2dcde0176129f4df49f13d4d3d6f7bea3023fb48 Mon Sep 17 00:00:00 2001
From: Jes Sorensen <Jes.Sorensen@redhat.com>
Date: Mon, 1 Sep 2014 09:25:04 -0400
Subject: [kernel] wait: Make the __wait_event*() interface more friendly

Message-id: <1409563584-9000-3-git-send-email-Jes.Sorensen@redhat.com>
Patchwork-id: 89938
O-Subject: [RHEL7.1 PATCH 02/82] wait: Make the __wait_event*() interface more friendly
Bugzilla: 1085530
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

From: Jes Sorensen <Jes.Sorensen@redhat.com>

Partial backport of the following upstream commit. Only changes to
___wait_event() have been included here:

commit 35a2af94c7ce7130ca292c68b1d27fcfdb648f6b
Author: Peter Zijlstra <peterz@infradead.org>
Date:   Wed Oct 2 11:22:33 2013 +0200

    sched/wait: Make the __wait_event*() interface more friendly

    Change all __wait_event*() implementations to match the corresponding
    wait_event*() signature for convenience.

    In particular this does away with the weird 'ret' logic. Since there
    are __wait_event*() users this requires we update them too.

    Reviewed-by: Oleg Nesterov <oleg@redhat.com>
    Signed-off-by: Peter Zijlstra <peterz@infradead.org>
    Link: http://lkml.kernel.org/r/20131002092529.042563462@infradead.org
    Signed-off-by: Ingo Molnar <mingo@kernel.org>

Signed-off-by: Jes Sorensen <Jes.Sorensen@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/include/linux/wait.h b/include/linux/wait.h
index d1a5883..3f49773 100644
--- a/include/linux/wait.h
+++ b/include/linux/wait.h
@@ -180,8 +180,9 @@ wait_queue_head_t *bit_waitqueue(void *, int);
 #define ___wait_nop_ret  int ret __always_unused
 
 #define ___wait_event(wq, condition, state, exclusive, ret, cmd) \
-do {         \
+({         \
  __label__ __out;      \
+ long __ret = ret;      \
  DEFINE_WAIT(__wait);      \
          \
  for (;;) {       \
@@ -194,7 +195,7 @@ do {         \
    break;      \
          \
   if (___wait_signal_pending(state)) {   \
-   ret = -ERESTARTSYS;    \
+   __ret = -ERESTARTSYS;    \
    if (exclusive) {    \
     abort_exclusive_wait(&wq, &__wait,  \
            state, NULL);  \
@@ -206,8 +207,8 @@ do {         \
   cmd;       \
  }        \
  finish_wait(&wq, &__wait);     \
-__out: ;        \
-} while (0)
+__out: __ret;        \
+})
 
 #define __wait_event(wq, condition)      \
 do {         \
-- 
1.7.1